src/createEpicMiddleware.ts

```diff
@@ -106,8 +132,37 @@ export function createEpicMiddleware<
         'epicMiddleware.run(rootEpic) called before the middleware has been setup by redux. Provide the epicMiddleware instance to createStore() first.'
       );
     }
+
+    Object.defineProperty(rootEpic, 'id', {
+      value: Math.random().toString(16),
+    });
+
     epic$.next(rootEpic);
   };
 
+  epicMiddleware.sigterm = (rootEpic) => {
+    const id = rootEpic.id;
+    const thread = threads[id];
+
+    setTimeout(() => {
+      store.dispatch({ type: getSigtermActionType(id) });
+    }, 0);
+
+    return thread.output$;
+  };
+
+  epicMiddleware.kill = (rootEpic) => {
+    const id = rootEpic.id;
+    const thread = threads[id];
+
+    if (!thread) { /* epic has never been run */ }
+
+    setTimeout(() => {
+      thread.kill$.next();
+    }, 0);
+
+    return thread.output$;
+  };
+
```

Comment:
### Preconditions

1. every running epic (thread) should be assigned a thread `id` to allow the middleware (thread manager) to terminate/kill threads individually

### Interface

1. **sigterm**: terminates the actions stream so that the thread would still have the time to complete any on-going job (e.g. async tasks), emit their last output and then exit gracefully
2. **kill**: terminates the thread directly, any on-going job would be abruptly killed straight away.

### Potential Usage

```typescript
const index = async (req, res) => {
  epicMiddleware.run(rootEpic);
  store.dispatch({ type: 'FETCH_TODOS' });
  
  const thread$ = epicMiddleware.sigterm(rootEpic);
  await thread$.toPromise();


  ReactDOMServer.renderToNodeStream(<Provider store={store}><App /></Provider>).pipe(res);
};
```

### Pros

1. Being able to manage every process individually (`run`, `kill`, `terminate`).

### Cons

1. need for extra `killall` and `termall` methods to kill and terminate all the currently running threads
2. if a rootEpic is run twice, then `kill` and `terminate` would affect all their instances (not necessarily a problem though)
3. `sigterm` can only terminate the action stream, so epics not propagating this completion would end up in a deadlock (as brilliantly highlighted by @liamzdenek [here](https://github.com/redux-observable/redux-observable/pull/539#issuecomment-543479927))

### Workarounds

1. Implement timeout to kill deadlocking processes

```typescript
const timeout = (middleware, rootEpic) => new Promise((resolve) => {
  setTimeout(() => {
    const thread$ = middleware.kill(rootEpic);
    thread$.toPromise().then(resolve);
  }, 2000);
});

const index = async (req, res) => {
  epicMiddleware.run(rootEpic);
  store.dispatch({ type: 'FETCH_TODOS' });
  
  const thread$ = epicMiddleware.sigterm(rootEpic);
  await Promise.race([
    thread$.toPromise(),
    timeout(epicMiddleware, rootEpic),
  ]);


  ReactDOMServer.renderToNodeStream(<Provider store={store}><App /></Provider>).pipe(res);
};
```

___

### Motivations

While SSR is probably one of the main use cases for implementing such API, it would also find applications in projects that heavily rely in code splitting. 

I am currently working on a project that mounts/unmounts reducers, components and epics depending on routing.

While we can easily add/remove reducers and components at runtime, it is impossible to do so with epics, since they cannot be terminated once ran.

**It really feels like `run` is missing its complementary api `terminate`**. (e.g. [redux#replaceReducer](https://redux.js.org/api/store#replacereducernextreducer))